# 7 Days of Code - Alura

#Importando bibliotecas a serem utilizadas

In [1]:
import pandas as pd

#Carregando os dados

In [2]:
# Etapa de carregar os dados referentes aos emprestimos de exemplares nas bibliotecas da UFRN.
#Os dados foram enviados em arquivos separados, portanto cada um deles será carregado em um dataframe distinto.

dados_2010_1 = pd.read_csv('/content/emprestimos-20101.csv')
dados_2010_2 = pd.read_csv('/content/emprestimos-20102.csv')
dados_2011_1 = pd.read_csv('/content/emprestimos-20111.csv')
dados_2011_2 = pd.read_csv('/content/emprestimos-20112.csv')
dados_2012_1 = pd.read_csv('/content/emprestimos-20121.csv')
dados_2012_2 = pd.read_csv('/content/emprestimos-20122.csv')
dados_2013_1 = pd.read_csv('/content/emprestimos-20131.csv')
dados_2013_2 = pd.read_csv('/content/emprestimos-20132.csv')
dados_2014_1 = pd.read_csv('/content/emprestimos-20141.csv')
dados_2014_2 = pd.read_csv('/content/emprestimos-20142.csv')
dados_2015_1 = pd.read_csv('/content/emprestimos-20151.csv')
dados_2015_2 = pd.read_csv('/content/emprestimos-20152.csv')
dados_2016_1 = pd.read_csv('/content/emprestimos-20161.csv')
dados_2016_2 = pd.read_csv('/content/emprestimos-20162.csv')
dados_2017_1 = pd.read_csv('/content/emprestimos-20171.csv')
dados_2017_2 = pd.read_csv('/content/emprestimos-20172.csv')
dados_2018_1 = pd.read_csv('/content/emprestimos-20181.csv')
dados_2018_2 = pd.read_csv('/content/emprestimos-20182.csv')
dados_2019_1 = pd.read_csv('/content/emprestimos-20191.csv')
dados_2019_2 = pd.read_csv('/content/emprestimos-20192.csv')
dados_2020_1 = pd.read_csv('/content/emprestimos-20201.csv')

In [3]:
#Importando os dados dos exemplares, também disponibilizado pela UFRN.

dados_exemplares = pd.read_parquet('/content/dados_exemplares.parquet')

Verificando se todos os dados possuem as mesmas colunas para então realizar o processo de concatenação dos dados. Para isso podemos utilizar o método da bibliteca pandas 'concat()', que apenas irá executar corretamente se todas as colunas forem iguais.

In [4]:
# Para facilitar a unificação dos dados, inicialmente vamos declarar uma lista de dataframes.

lista_de_dados = [dados_2010_1, dados_2010_2, dados_2011_1, dados_2011_2, dados_2012_1, dados_2012_2, dados_2013_1, dados_2013_2,
                  dados_2014_1, dados_2014_2, dados_2015_1, dados_2015_2, dados_2016_1, dados_2016_2, dados_2017_1, dados_2017_2,
                  dados_2018_1, dados_2018_2, dados_2019_1, dados_2019_2, dados_2020_1]
# Em seguida, podemos adicionar esta lista de dataframes dentro da função concat da biblioteca pandas, para então gerar um único dataframe com todos os dados de emprestimos.

dados_emprestimos = pd.concat(lista_de_dados, ignore_index=True)
dados_emprestimos.head()   #verificar quais são as colunas e como os dados estão carregados.

id_emprestimo codigo_barras                 data_renovacao  \
0         709684       L095049                            NaN   
1         709685       L167050                            NaN   
2         709686    2006017618  2010/01/26 08:07:01.738000000   
3         709687       L184117  2010/01/18 11:07:46.470000000   
4         709684       L095049                            NaN   

                 data_emprestimo                 data_devolucao  \
0  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   
1  2010/01/04 07:44:10.750000000  2010/01/12 07:34:13.934000000   
2  2010/01/04 08:08:44.081000000  2010/02/25 07:36:25.800000000   
3  2010/01/04 08:24:21.284000000  2010/02/03 08:58:45.692000000   
4  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   

   matricula_ou_siape    tipo_vinculo_usuario  
0        2.008023e+09      ALUNO DE GRADUAÇÃO  
1        2.008023e+09      ALUNO DE GRADUAÇÃO  
2        2.008112e+09  ALUNO DE PÓS-GRADUAÇÃO  
3        2.007211e+08      ALUNO DE GRADUAÇÃO  
4        2.008023e+09      ALUNO DE GRADUAÇÃO

In [5]:
# Verificando quantas entradas há em nosso dataframe.
dados_emprestimos.shape

(2258018, 7)

In [6]:
# Verificando o tipo de dados de cada uma das colunas.
dados_emprestimos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2258018 entries, 0 to 2258017
Data columns (total 7 columns):
 #   Column                Dtype  
---  ------                -----  
 0   id_emprestimo         int64  
 1   codigo_barras         object 
 2   data_renovacao        object 
 3   data_emprestimo       object 
 4   data_devolucao        object 
 5   matricula_ou_siape    float64
 6   tipo_vinculo_usuario  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 120.6+ MB


Os dados de emprestimos 'raw' fornecidos para este desafio não estão limpos, sendo assim é necessário dar início ao processamento destes dados. Este processo seguirá as etapas abaixo:

* Verificar dados em duplicidade.
* Identificar elementos nulos.
* Identificar os tipos de entradas para cada coluna e se necessário alterá-los.



In [7]:
# esta etapa tem a função de verificar se há dados duplicados no nosso dataframe.

dados_emprestimos.value_counts()

id_emprestimo  codigo_barras  data_renovacao                 data_emprestimo                data_devolucao                 matricula_ou_siape  tipo_vinculo_usuario  
709704         2006021442     2010/01/17 18:56:43.923000000  2010/01/04 09:58:46.522000000  2010/02/01 13:59:02.528000000  2.008010e+09        ALUNO DE GRADUAÇÃO        7
709705         2009032168     2010/01/17 18:56:43.909000000  2010/01/04 09:58:46.549000000  2010/01/20 13:14:34.464000000  2.008010e+09        ALUNO DE GRADUAÇÃO        7
709698         2009047725     2010/01/18 14:44:41.163000000  2010/01/04 09:21:19.099000000  2010/02/02 12:02:38.444000000  2.009047e+09        ALUNO DE GRADUAÇÃO        3
709686         2006017618     2010/01/26 08:07:01.738000000  2010/01/04 08:08:44.081000000  2010/02/25 07:36:25.800000000  2.008112e+09        ALUNO DE PÓS-GRADUAÇÃO    2
2461099        2012008348     2016/06/27 20:35:46.855000000  2016/06/14 11:20:33.254000000  2016/07/11 14:37:29.000000000  2.015022e+09        ALUNO D

Como foi possível verificar acima, existem entradas nos dados de emprestimos que são duplicados, cuja contagem chega em 7. Sendo assim, precisamos remover os dados duplicado, para que as futuras análises sejam fiéis a realidade do sistema de bibliotecas aos quais os dados pertencem.

In [8]:
# A biblioteca pandas possui um método que permite remover facilmente entradas duplicadas em nosso dataframe.
# O método em questão é o .drop_duplicate()

dados_emprestimos = dados_emprestimos.drop_duplicates()

In [9]:
# Verificando se de fato os elementos duplicados foram removidos.

dados_emprestimos.value_counts()

id_emprestimo  codigo_barras  data_renovacao                 data_emprestimo                data_devolucao                 matricula_ou_siape  tipo_vinculo_usuario  
709686         2006017618     2010/01/26 08:07:01.738000000  2010/01/04 08:08:44.081000000  2010/02/25 07:36:25.800000000  2.008112e+09        ALUNO DE PÓS-GRADUAÇÃO    1
2336122        2011033079     2016/11/21 02:01:58.488000000  2016/11/10 17:00:12.989000000  2016/12/02 14:21:49.000000000  2.016023e+09        ALUNO DE GRADUAÇÃO        1
2336103        2014085936     2016/11/25 19:25:36.215000000  2016/11/10 16:54:52.447000000  2016/12/13 14:30:52.000000000  2.014047e+09        ALUNO DE GRADUAÇÃO        1
2336104        2011027960     2016/11/23 17:27:54.642000000  2016/11/10 16:55:17.956000000  2016/11/28 10:32:38.000000000  2.016083e+09        ALUNO DE GRADUAÇÃO        1
2336105        2014000897     2016/12/02 15:02:53.356000000  2016/11/10 16:55:19.851000000  2016/12/23 13:53:23.000000000  2.014071e+09        ALUNO D

In [10]:
# Após confirmar que não há mais dados duplicados, podemos ferificar se há dados nulos no nosso dataframe.
# Para isso, podemos utilizar os métodos da biblioteca pandas, isnull e sum(), para identificar elementos nulos e quantifica-los.

dados_emprestimos.isnull().sum()

id_emprestimo                 0
codigo_barras                 0
data_renovacao          1285709
data_emprestimo               0
data_devolucao             6471
matricula_ou_siape         3170
tipo_vinculo_usuario          0
dtype: int64

As colunas data_renovação, data_devolução e matricula_ou_siape apresentam valores nulos.

Dentre estas, a única que potencialmente representa um problema para as análises que se seguirão é a matricula_ou_siape, sendo assim, os registros nulos associados a ela serão removidos.

A coluna data_renovação possui entradas nulas, pois nem todo emprestimo de livro implica na renovação do emprestimo.

A coluna data_devolução com valores nulos, em um primeiro momento pode levar a conclusão de ser uma entrada com erro, no entando há casos em que o usuário(a) retira um exemplar emprestado e não devolve no prazo e tão pouco realiza a renovação do emprestimo do mesmo.

In [11]:
# Procedimento para identificar o index dos registro de matrícula_ou_siape.
# Após a identificação, iremos utilizar estes index para remover os registros correspondentes.

index_a_serem_removidos = dados_emprestimos[['matricula_ou_siape']].isnull().query('matricula_ou_siape == True').index
dados_emprestimos_V2 = dados_emprestimos.drop(index_a_serem_removidos)

In [12]:
# Procedimento para confirmar se há valores nulos remanecentes na coluna matricula_ou_siape.
dados_emprestimos_V2[['matricula_ou_siape']].isnull().value_counts()

matricula_ou_siape
False                 2254811
dtype: int64

Após remover diversas entradas do nosso DataFrame, os números dos índice ficam despadronizados, o que pode levar a erros ou mau funcionamento de códigos. Portanto, é uma boa prática, re-iniciar a contagem do índice, utilizando o método da biblioteca pandas conhecido como reset_index().

Vale destacar que é necessário inserir o argumento drop = True para garantir que o índice original será descartado e substituído pelo novo índice.

In [13]:
dados_emprestimos_V2.reset_index(drop = True, inplace = True)
dados_emprestimos_V2.head()

id_emprestimo codigo_barras                 data_renovacao  \
0         709684       L095049                            NaN   
1         709685       L167050                            NaN   
2         709686    2006017618  2010/01/26 08:07:01.738000000   
3         709687       L184117  2010/01/18 11:07:46.470000000   
4         709688    2007027500  2010/01/19 15:59:26.464000000   

                 data_emprestimo                 data_devolucao  \
0  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   
1  2010/01/04 07:44:10.750000000  2010/01/12 07:34:13.934000000   
2  2010/01/04 08:08:44.081000000  2010/02/25 07:36:25.800000000   
3  2010/01/04 08:24:21.284000000  2010/02/03 08:58:45.692000000   
4  2010/01/04 08:27:33.569000000  2010/02/03 13:06:30.662000000   

   matricula_ou_siape    tipo_vinculo_usuario  
0        2.008023e+09      ALUNO DE GRADUAÇÃO  
1        2.008023e+09      ALUNO DE GRADUAÇÃO  
2        2.008112e+09  ALUNO DE PÓS-GRADUAÇÃO  
3        2.007211e+08      ALUNO DE GRADUAÇÃO  
4        2.009046e+09      ALUNO DE GRADUAÇÃO

Agora que o dataframe esta corrigido, iremos ajustar os tipos de entradas para cada coluna.

In [14]:
# A matricula_ou_siape esta com tipo de dado igual a float, e por isso temos uma visualização prejudicada destes valores.
# Não é esperado realizar nenhuma operação matemática com estes dados, pois ele são tratados como variáveis categóricas, portanto podemos mudar o seu tipo para str.

dados_emprestimos_V2[['matricula_ou_siape']] = dados_emprestimos_V2[['matricula_ou_siape']].astype(str)

In [15]:
dados_emprestimos_V2.head()

id_emprestimo codigo_barras                 data_renovacao  \
0         709684       L095049                            NaN   
1         709685       L167050                            NaN   
2         709686    2006017618  2010/01/26 08:07:01.738000000   
3         709687       L184117  2010/01/18 11:07:46.470000000   
4         709688    2007027500  2010/01/19 15:59:26.464000000   

                 data_emprestimo                 data_devolucao  \
0  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   
1  2010/01/04 07:44:10.750000000  2010/01/12 07:34:13.934000000   
2  2010/01/04 08:08:44.081000000  2010/02/25 07:36:25.800000000   
3  2010/01/04 08:24:21.284000000  2010/02/03 08:58:45.692000000   
4  2010/01/04 08:27:33.569000000  2010/02/03 13:06:30.662000000   

  matricula_ou_siape    tipo_vinculo_usuario  
0       2008023265.0      ALUNO DE GRADUAÇÃO  
1       2008023265.0      ALUNO DE GRADUAÇÃO  
2       2008112114.0  ALUNO DE PÓS-GRADUAÇÃO  
3        200721119.0      ALUNO DE GRADUAÇÃO  
4       2009045971.0      ALUNO DE GRADUAÇÃO

Agora podemos seguir o mesmo procedimento de limpeza dos dados para o DataFrame dados_exemplares.

In [16]:
dados_exemplares.head()

id_exemplar codigo_barras            colecao  \
index                                                 
0                5       L000003  Acervo Circulante   
1                4       L000002  Acervo Circulante   
2                3       L000001  Acervo Circulante   
3                7       L000114  Acervo Circulante   
5               10       L000041  Acervo Circulante   

                           biblioteca status_material  localizacao  \
index                                                                
0      Biblioteca Central Zila Mamede         REGULAR          694   
1      Biblioteca Central Zila Mamede         REGULAR          688   
2      Biblioteca Central Zila Mamede        ESPECIAL          638   
3      Biblioteca Central Zila Mamede         REGULAR          616   
5      Biblioteca Central Zila Mamede        ESPECIAL          657   

       registro_sistema  
index                    
0                     1  
1                     1  
2                     1  
3                     5  
5                    15

In [17]:
dados_exemplares.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 546237 entries, 0 to 568584
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id_exemplar       546237 non-null  int64 
 1   codigo_barras     546237 non-null  object
 2   colecao           546237 non-null  object
 3   biblioteca        546237 non-null  object
 4   status_material   546237 non-null  object
 5   localizacao       546237 non-null  int64 
 6   registro_sistema  546237 non-null  int64 
dtypes: int64(3), object(4)
memory usage: 33.3+ MB


In [18]:
dados_exemplares.isnull().sum()

id_exemplar         0
codigo_barras       0
colecao             0
biblioteca          0
status_material     0
localizacao         0
registro_sistema    0
dtype: int64

Como observado nos comandos acima, o DataFrame dados_exemplares, possui 7 colunas, 546237 registros e nanhum dado nulo. Contudo a contagem dos índices, não esta correta, então a proxima etapa consiste em re-iniciar a contagem do índice.

In [19]:
dados_exemplares.reset_index(drop = True, inplace = True)
dados_exemplares.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 546237 entries, 0 to 546236
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id_exemplar       546237 non-null  int64 
 1   codigo_barras     546237 non-null  object
 2   colecao           546237 non-null  object
 3   biblioteca        546237 non-null  object
 4   status_material   546237 non-null  object
 5   localizacao       546237 non-null  int64 
 6   registro_sistema  546237 non-null  int64 
dtypes: int64(3), object(4)
memory usage: 29.2+ MB


Após corrigir inconsistencias nos dos DataFrames desta análise, podemos mescla-los para facilitar futuras análises. Para isso podemos utilizar o comando da biblioteca pandas merge(), que trabalha de maneira semelhando ao JOIN do SQL. Este método vai estipular automaticamente ou não uma coluna de chave primária, neste caso a coluna código de barras, e adicionar os valores dos dois DataFrames que são correspondentes.

In [20]:
dados_completos = dados_emprestimos_V2.merge(dados_exemplares)

In [21]:
dados_completos.head()

id_emprestimo codigo_barras                 data_renovacao  \
0         709684       L095049                            NaN   
1         726965       L095049  2010/03/31 12:41:08.684000000   
2         709685       L167050                            NaN   
3         717763       L167050                            NaN   
4         729996       L167050  2010/03/17 09:08:26.874000000   

                 data_emprestimo                 data_devolucao  \
0  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   
1  2010/03/01 12:40:56.212000000  2010/04/30 14:42:35.366000000   
2  2010/01/04 07:44:10.750000000  2010/01/12 07:34:13.934000000   
3  2010/02/11 11:16:30.335000000  2010/03/04 08:54:32.262000000   
4  2010/03/04 13:36:26.599000000  2010/03/31 18:04:29.669000000   

  matricula_ou_siape    tipo_vinculo_usuario  id_exemplar            colecao  \
0       2008023265.0      ALUNO DE GRADUAÇÃO        13259  Acervo Circulante   
1       2009121584.0  ALUNO DE PÓS-GRADUAÇÃO        13259  Acervo Circulante   
2       2008023265.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   
3       2008050352.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   
4        200722565.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   

                       biblioteca status_material  localizacao  \
0  Biblioteca Central Zila Mamede         REGULAR          630   
1  Biblioteca Central Zila Mamede         REGULAR          630   
2  Biblioteca Central Zila Mamede         REGULAR          647   
3  Biblioteca Central Zila Mamede         REGULAR          647   
4  Biblioteca Central Zila Mamede         REGULAR          647   

   registro_sistema  
0              4225  
1              4225  
2             25009  
3             25009  
4             25009

In [22]:
dados_completos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2069629 entries, 0 to 2069628
Data columns (total 13 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   id_emprestimo         int64 
 1   codigo_barras         object
 2   data_renovacao        object
 3   data_emprestimo       object
 4   data_devolucao        object
 5   matricula_ou_siape    object
 6   tipo_vinculo_usuario  object
 7   id_exemplar           int64 
 8   colecao               object
 9   biblioteca            object
 10  status_material       object
 11  localizacao           int64 
 12  registro_sistema      int64 
dtypes: int64(4), object(9)
memory usage: 221.1+ MB


Nos foi informado que os valores presentes na coluna localização, são referêntes ao gênero do exemplar. Portando vamos adicionar uma nova coluna que contenha de maneira evidente a qual gênero cada um dos exemplares pertence.


In [23]:
Genero = []
for CDU in dados_completos['localizacao']:
  if CDU <= 99:
    Genero.append('Generalidades, Ciência e Conhecimento')
  elif CDU <= 199:
    Genero.append('Filosofia e Psicologia')
  elif CDU <= 299:
    Genero.append('Religião')
  elif CDU <= 399:
    Genero.append('Ciências Sociais')
  elif CDU <= 499:
    Genero.append('Classe vaga. Provisoriamente não ocupada')
  elif CDU <= 599:
    Genero.append('Matemática e Ciências naturais')
  elif CDU <= 699:
    Genero.append('Ciências Aplicadas')
  elif CDU <= 799:
    Genero.append('Belas Artes')
  elif CDU <= 899:
    Genero.append('Linguagem, Lingua, Linguistica')
  else:
    Genero.append('Geoografia, Biografia, História')

dados_completos['Genero'] = Genero

dados_completos.head()

id_emprestimo codigo_barras                 data_renovacao  \
0         709684       L095049                            NaN   
1         726965       L095049  2010/03/31 12:41:08.684000000   
2         709685       L167050                            NaN   
3         717763       L167050                            NaN   
4         729996       L167050  2010/03/17 09:08:26.874000000   

                 data_emprestimo                 data_devolucao  \
0  2010/01/04 07:44:10.721000000  2010/01/05 16:26:12.662000000   
1  2010/03/01 12:40:56.212000000  2010/04/30 14:42:35.366000000   
2  2010/01/04 07:44:10.750000000  2010/01/12 07:34:13.934000000   
3  2010/02/11 11:16:30.335000000  2010/03/04 08:54:32.262000000   
4  2010/03/04 13:36:26.599000000  2010/03/31 18:04:29.669000000   

  matricula_ou_siape    tipo_vinculo_usuario  id_exemplar            colecao  \
0       2008023265.0      ALUNO DE GRADUAÇÃO        13259  Acervo Circulante   
1       2009121584.0  ALUNO DE PÓS-GRADUAÇÃO        13259  Acervo Circulante   
2       2008023265.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   
3       2008050352.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   
4        200722565.0      ALUNO DE GRADUAÇÃO        70865  Acervo Circulante   

                       biblioteca status_material  localizacao  \
0  Biblioteca Central Zila Mamede         REGULAR          630   
1  Biblioteca Central Zila Mamede         REGULAR          630   
2  Biblioteca Central Zila Mamede         REGULAR          647   
3  Biblioteca Central Zila Mamede         REGULAR          647   
4  Biblioteca Central Zila Mamede         REGULAR          647   

   registro_sistema              Genero  
0              4225  Ciências Aplicadas  
1              4225  Ciências Aplicadas  
2             25009  Ciências Aplicadas  
3             25009  Ciências Aplicadas  
4             25009  Ciências Aplicadas

Para a nossa análise, a coluna registro_sistema não possui valor, portanto podemos excluí-la.

In [24]:
dados_completos.drop(columns = ['registro_sistema'], axis = 1, inplace=True)

In [27]:
dados_completos.to_csv('dados_completos.csv', index = False)